File: src/Avalonia.Samples/CompleteApps/Avalonia.MusicStore/Views/MainWindow.axaml.cs
using System;
using Avalonia.Controls;
using Avalonia.Controls.Notifications;
using Avalonia.MusicStore.Messages;
using Avalonia.MusicStore.ViewModels;
using CommunityToolkit.Mvvm.Messaging;

namespace Avalonia.MusicStore.Views
{
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();

            // The designer cannot open a new window, so we don't register the messenger if we have a design session. 
            if (Design.IsDesignMode)
                return;

            // Whenever 'Send(new PurchaseAlbumMessage())' is called, invoke this callback on the MainWindow instance:
            WeakReferenceMessenger.Default.Register<MainWindow, PurchaseAlbumMessage>(this, static (w, m) =>
            {
                // Create an instance of MusicStoreWindow and set MusicStoreViewModel as its DataContext.
                var dialog = new MusicStoreWindow
                {
                    DataContext = new MusicStoreViewModel()
                };

                // Show dialog window and reply with returned AlbumViewModel or null when the dialog is closed.
                m.Reply(dialog.ShowDialog<AlbumViewModel?>(w));
            });
        }
    }
}

File: src/Avalonia.Samples/FSharp/CompleteApps/MusicStore/Views/MainWindow.axaml.fs
namespace MusicStore.Views

open Avalonia
open Avalonia.Controls
open Avalonia.Markup.Xaml
open CommunityToolkit.Mvvm.Messaging
open MusicStore.Messages
open System.Threading.Tasks
open MusicStore.ViewModels

type MainWindow() as this =
    inherit Window()

    let messenger =
        fun (w: MainWindow) (message: PurchaseAlbumMessage.Self) ->
            // Open the music store dialog and reply with the selected album (if any)
            let dialog = MusicStoreWindow(DataContext = MusicStoreViewModel())

            let response: Task<AlbumViewModel option> =
                task {
                    // Show the dialog modally relative to this window and await the selected album
                    let! selected = dialog.ShowDialog<AlbumViewModel>(w)
                    // If the window was closed without a selection, selected might be null
                    return
                        if obj.ReferenceEquals(selected, null) then
                            None
                        else
                            Some selected
                }

            message.Reply(response)

    do
        this.InitializeComponent()

        // Only register messenger handlers at runtime to avoid issues in the designer
        if not Design.IsDesignMode then
            // Register an AsyncRequestMessage handler that replies with a result
            WeakReferenceMessenger.Default.Register(this, messenger)

        // Kick off data loading once the window is opened and DataContext is ready
        this.Opened.Add(fun _ ->
            match this.DataContext with
            | :? MainWindowViewModel as vm -> vm.InitializeAsync() |> ignore
            | _ -> ()
        )

    member private this.InitializeComponent() =
#if DEBUG
        this.AttachDevTools()
#endif
        AvaloniaXamlLoader.Load(this)
